### ISSUE TITLE
Understanding the causal attention mask

### ISSUE THREAD

**Issue by SchweitzerGAO** [OpenGVLab/SDLM#4]
Hi there, 
I wonder why you used a standard causal mask at the top-left position of the attention mask

<img width="1340" height="1440" alt="Image" src="https://github.com/user-attachments/assets/a094b130-567e-49eb-abdc-f0c6d6e402e6" />

Is block-wise causal attention feasible(like Block Diffusion[1] does) with your settings?

<img width="1246" height="1146" alt="Image" src="https://github.com/user-attachments/assets/a08f48f1-0983-41db-9d44-34d87513e8e6" />

Thank you!

[1] Block Diffusion: Interpolating Between Autoregressive and Diffusion Language Models: https://arxiv.org/abs/2503.09573

**Comment by SchweitzerGAO** [comment]
Thank you for your quick reply! Did you mean that a block-wise causal mask is less effective in transforming AR to MDMs? is there any ablations? Or is the block-wise one just not feasible in your work?

**Comment by yuecao0119** [comment]
Our approach, using block-wise attention for historical information, still outperforms existing diffusion models like Dream and LLaDA. However, compared to causal attention, the speed improvement is slightly slower. This is due to our goal of keeping training costs low, using only 3.5M samples.

**Comment by Knlife** [comment]
Thank you for your reply. 
Both the paper and your previous response focus on the attention mask **inside the blocks** (the right-bottom protion of the Figure2(a) in the paper), but not the design of **attention mask in the history tokens** (the left-top protion of the Figure2(a), i.e. the green block in the first figure of the [issue](https://github.com/OpenGVLab/SDLM/issues/4#issue-3540505969). Could you please indicate the section of your paper that discusses the history attention mask in detail?

**Comment by yuecao0119** [comment]
Sorry, I misunderstood your question. We didn't conduct a complete ablation experiment in the final experiment. We only conducted experiments during the initial exploration of the method, trained for only one epoch, and the data used differs from the final data. However, we can provide reference results, which are as follows:

|Model|math-500|GSM8k|GPQA|human-eval|MBPP|
|---|---|---|---|---|---|
|BD3LM-D4|49.8|69.75|15.66|59.15|38.2|
|SDLM-D4|49.4|71.34|20.2|59.76|47.2|

It is worth noting that we found that the MBPP results under block attention are poor because some data are output in the wrong format such as [BBEGIN] when output, while the correct format is [BEGIN].
